api/src/services/script-runner.service.ts

```diff
@@ -28,34 +31,543 @@ export class ScriptRunnerService {
+    return {
+      createdAt: dbAddress[0]['created_at'],
+      placeName: dbAddress[0]['place_name'],
+      city: dbAddress[0]['city'],
+      county: dbAddress[0]['county'] || jurisdictionName,
+      state: dbAddress[0]['state'],
+      street: dbAddress[0]['street'],
+      street2: dbAddress[0]['street2'],
+      zipCode: dbAddress[0]['zip_code'],
+      latitude: dbAddress[0]['latitude'],
+      longitude: dbAddress[0]['longitude'],
+    };
+  }
+
+  /**
+   *
+   * @param req incoming request object
+   * @param dataTransferDTO data transfer endpoint args. Should contain foreign db connection string
+   * @returns successDTO
+   * @description transfers data from foreign data into the database this api normally connects to
+   */
+  async transferJurisdictionListingData(
+    req: ExpressRequest,
+    dataTransferDTO: DataTransferDTO,
+    prisma?: PrismaClient,
+  ): Promise<SuccessDTO> {
+    // script runner standard start up
+    const requestingUser = mapTo(User, req['user']);
+    await this.markScriptAsRunStart(
+      `data transfer listings ${dataTransferDTO.jurisdiction}`,
+      requestingUser,
+    );
+
+    // connect to foreign db based on incoming connection string
+    const client =
+      prisma ||
+      new PrismaClient({
+        datasources: {
+          db: {
+            url: dataTransferDTO.connectionString,
+          },
+        },
+      });
     await client.$connect();
 
-    // get data
-    const res =
-      await client.$queryRaw`SELECT id, name FROM jurisdictions WHERE name = 'San Mateo'`;
-    console.log(res);
+    const doorwayJurisdiction = await this.prisma.jurisdictions.findFirst({
+      where: { name: dataTransferDTO.jurisdiction },
+    });
+
+    if (!doorwayJurisdiction) {
+      throw new Error(
+        `${dataTransferDTO.jurisdiction} county doesn't exist in Doorway database`,
+      );
+    }
+
+    // get jurisdiction
+    const jurisdiction: { id: string }[] =
+      await client.$queryRaw`SELECT id, name FROM jurisdictions WHERE name = ${dataTransferDTO.jurisdiction}`;
+
+    if (!jurisdiction) {
+      throw new Error(
+        `${dataTransferDTO.jurisdiction} county doesn't exist in foreign database`,
+      );
+    }
+
+    const listingQuery = `SELECT * FROM listings WHERE jurisdiction_id = '${jurisdiction[0].id}'`;
+    const listings: any[] = await client.$queryRawUnsafe(listingQuery);
+
+    if (listings?.length) {
+      const priorityTypes: { id: string; name: string }[] =
+        await client.$queryRaw`SELECT * FROM unit_accessibility_priority_types`;
+      const doorwayPriorityTypes =
+        await this.prisma.unitAccessibilityPriorityTypes.findMany();
+      const rentTypes: { id: string; name: string }[] =
+        await client.$queryRaw`SELECT * FROM unit_types`;
+      const doorwayRentTypes = await this.prisma.unitRentTypes.findMany();
+      console.log(`migrating ${listings.length} listings`);
+      listings.forEach(async (listing) => {
```

Comment:
`.forEach()` doesn't play nicely with async as it currently stands this is firing off async calls and not waiting for each of them properly I don't think https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop


You probs wanna switch this to a different method of looping over the listings

Reply:
Fixed. I switched it to for loops. I didn't know that the forEach was an issue as it was working for me, but my guess is my data set is too small to run into the issue so good catch!